File: web/static/macro-import.js
// Macro import: a macro file written for an installed terminal emulator,
// translated into a recording this terminal can replay.
//
// The report is the feature. A shop moving off a desktop client has years of
// macros, and the question that decides whether it can move is not "did this
// one convert" but "which lines did not, and what do I do about them". So the
// panel leads with the steps it produced, then lists every line it would not
// translate with the reason, and only then offers to use the result — with
// Debug named as the way to run it the first time, because a translated flow
// deserves to be stepped through before it types into a live host.
//
// The server has already loaded the recording into the session by the time
// this renders, so closing the panel reloads the page: the Automation menu
// has to come back showing the loaded recording rather than the import button
// that has just been used.
(function () {
  "use strict";

  document.addEventListener("DOMContentLoaded", function () {
    var modal = document.querySelector("[data-macro-import-modal]");
    var fileInput = document.querySelector("[data-macro-import-file]");
    if (!modal || !fileInput) {
      return;
    }

    var statusEl = modal.querySelector("[data-macro-import-status]");
    var summaryEl = modal.querySelector("[data-macro-import-summary]");
    var advisoriesEl = modal.querySelector("[data-macro-import-advisories]");
    var notesWrap = modal.querySelector("[data-macro-import-notes-wrap]");
    var notesEl = modal.querySelector("[data-macro-import-notes]");
    var loaded = false;

    function setStatus(text, isError) {
      if (!statusEl) {
        return;
      }
      statusEl.textContent = text || "";
      statusEl.hidden = !text;
      if (isError) {
        statusEl.setAttribute("data-error", "1");
      } else {
        statusEl.removeAttribute("data-error");
      }
    }

    function clear() {
      summaryEl.textContent = "";
      advisoriesEl.textContent = "";
      advisoriesEl.hidden = true;
      notesEl.textContent = "";
      notesWrap.hidden = true;
      setStatus("", false);
    }

    function plural(count, one, many) {
      return count === 1 ? one : many;
    }

    // renderReport shows the same account whether or not anything loaded: a
    // file that produced no steps still has a reason per line, and that is
    // the answer somebody needs.
    function renderReport(report, wasLoaded) {
      var notes = report.notes || [];
      var steps = report.stepTotal || 0;
      var lines = [];

      if (wasLoaded) {
        lines.push(
          "Loaded " +
            report.name +
            " — " +
            steps +
            " " +
            plural(steps, "step", "steps") +
            " from " +
            report.translated +
            " of " +
            report.statements +
            " " +
            plural(report.statements, "statement", "statements") +
            "."
        );
      } else {
        lines.push(
          "Nothing in this file became a step. " +
            report.statements +
            " " +
            plural(report.statements, "statement", "statements") +
            " were read."
        );
      }
      if (notes.length) {
        lines.push(
          notes.length +
            " " +
            plural(notes.length, "line needs", "lines need") +
            " a person: they are listed below and were left out of the recording."
        );
      } else if (wasLoaded) {
        lines.push("Every statement in the file translated.");
      }
      summaryEl.textContent = lines.join(" ");

      var advisories = report.advisories || [];
      advisoriesEl.textContent = "";
      if (advisories.length) {
        for (var a = 0; a < advisories.length; a++) {
          var item = document.createElement("li");
          item.textContent = advisories[a];
          advisoriesEl.appendChild(item);
        }
      }
      advisoriesEl.hidden = advisories.length === 0;

      notesEl.textContent = "";
      for (var i = 0; i < notes.length; i++) {
        var note = notes[i];
        var item = document.createElement("li");

        var head = document.createElement("div");
        head.className = "macro-import-note-head";
        var lineLabel = document.createElement("span");
        lineLabel.className = "macro-import-line";
        lineLabel.textContent = "Line " + note.line;
        head.appendChild(lineLabel);
        var code = document.createElement("code");
        code.textContent = note.source;
        head.appendChild(code);
        item.appendChild(head);

        var reason = document.createElement("p");
        reason.className = "macro-import-note-reason";
        reason.textContent = note.reason;
        item.appendChild(reason);

        notesEl.appendChild(item);
      }
      notesWrap.hidden = notes.length === 0;

      if (wasLoaded) {
        setStatus(
          "Step through it with Debug recording before you play it: a translated flow has never met this host.",
          false
        );
      }
    }

    function closeModal() {
      modal.hidden = true;
      if (window.ThreeSeventyWeb && window.ThreeSeventyWeb.popModal) {
        window.ThreeSeventyWeb.popModal(modal);
      }
      if (loaded) {
        // The session now holds a recording the menu does not know about yet.
        window.location.reload();
      }
    }

    function openModal() {
      modal.hidden = false;
      if (window.ThreeSeventyWeb && window.ThreeSeventyWeb.pushModal) {
        window.ThreeSeventyWeb.pushModal(modal, closeModal);
      }
    }

    function importFile(file) {
      loaded = false;
      clear();
      openModal();
      setStatus("Translating " + file.name + "…", false);

      var body = new FormData();
      body.append("macro", file);
      fetch("/workflow/import-macro", {
        method: "POST",
        headers: { Accept: "application/json" },
        body: body
      })
        .then(function (res) {
          return res
            .json()
            .catch(function () {
              return {};
            })
            .then(function (payload) {
              return { ok: res.ok, status: res.status, payload: payload };
            });
        })
        .then(function (result) {
          if (result.ok) {
            loaded = true;
            renderReport(result.payload, true);
            return;
          }
          if (result.payload && result.payload.report) {
            renderReport(result.payload.report, false);
            setStatus(result.payload.error || "This file produced no steps.", true);
            return;
          }
          setStatus(
            (result.payload && result.payload.error) || "The macro file could not be imported.",
            true
          );
        })
        .catch(function () {
          setStatus("The macro file could not be imported.", true);
        });
    }

    var openButtons = document.querySelectorAll("[data-macro-import-open]");
    for (var i = 0; i < openButtons.length; i++) {
      openButtons[i].addEventListener("click", function () {
        fileInput.click();
      });
    }

    fileInput.addEventListener("change", function () {
      if (fileInput.files && fileInput.files.length > 0) {
        importFile(fileInput.files[0]);
        // Let the same file be picked again after a failed attempt.
        fileInput.value = "";
      }
    });

    var closeButtons = modal.querySelectorAll("[data-macro-import-close]");
    for (var k = 0; k < closeButtons.length; k++) {
      closeButtons[k].addEventListener("click", closeModal);
    }

    modal.addEventListener("click", function (event) {
      if (event.target === modal) {
        closeModal();
      }
    });
  });
})();

File: cmd/3270Web/macroimport.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"path"
	"path/filepath"
	"strings"

	"github.com/gin-gonic/gin"

	"github.com/jnnngs/3270Web/internal/macroimport"
	"github.com/jnnngs/3270Web/internal/session"
)

// Macro import. A shop with hundreds of recorded macros has hundreds of
// reasons not to move, and rewriting them by hand is the reason an evaluation
// stops. These two endpoints read a macro file and hand back the recording it
// becomes, along with a line-by-line report of what did not translate —
// see internal/macroimport for what the translator will and will not do.
//
// There are two of them because there are two jobs. The browser endpoint
// loads the result into the session, so the operator can step through it
// against the host on the next click; the API endpoint translates and returns
// without touching any session, so a script can walk a directory of three
// hundred files in one pass and get three hundred recordings and three
// hundred reports back.

// A macro file is source text somebody wrote. The recording upload limit is
// two megabytes; a quarter of that is far past any hand-written macro and
// still bounded well below what a request should carry.
const maxMacroUploadBytes = 512 * 1024

// macroTranslationResponse is what both endpoints answer with. The report is
// the part that matters: steps alone would be the silent partial conversion
// this feature exists to avoid.
type macroTranslationResponse struct {
	Name       string             `json:"name"`
	StepTotal  int                `json:"stepTotal"`
	Statements int                `json:"statements"`
	Translated int                `json:"translated"`
	Notes      []macroimport.Note `json:"notes"`
	Advisories []string           `json:"advisories,omitempty"`
	Workflow   *WorkflowConfig    `json:"workflow,omitempty"`
}

// ImportMacroHandler translates an uploaded macro file and loads the result
// as this session's recording, so Play and Debug reach it immediately.
//
// The report comes back in the response rather than in a file: the moment to
// read which twelve lines need a human is before the flow is played against a
// host, not after.
func (app *App) ImportMacroHandler(c *gin.Context) {
	s := app.getSession(c)
	if s == nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "session not found"})
		return
	}
	playing := false
	withSessionLock(s, func() {
		playing = s.Playback != nil && s.Playback.Active
	})
	if playing {
		c.JSON(http.StatusConflict, gin.H{"error": "cannot load a recording during playback"})
		return
	}

	name, source, err := macroFileFromRequest(c)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	result := macroimport.Translate(source)
	host, port := workflowTargetForSession(s)
	workflow := macroWorkflow(host, port, name, result)

	if len(result.Steps) == 0 {
		// Nothing to load, but the report is the answer to "why not" and is
		// worth more than the status code on its own.
		c.JSON(http.StatusUnprocessableEntity, gin.H{
			"error": "no line of this file translated into a step",
			"report": macroTranslationResponse{
				Name:       recordingNameForMacro(name),
				Statements: result.Statements,
				Translated: result.Translated,
				Notes:      result.Notes,
				Advisories: result.Advisories,
			},
		})
		return
	}

	if err := macroBlocksResolve(workflow); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	raw, err := json.Marshal(workflow)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "the translated recording could not be encoded"})
		return
	}
	recordingName := recordingNameForMacro(name)
	storeLoadedWorkflow(s, recordingName, raw, len(workflow.Steps), prettyWorkflowPayload(raw))

	c.JSON(http.StatusOK, macroTranslationResponse{
		Name:       recordingName,
		StepTotal:  len(workflow.Steps),
		Statements: result.Statements,
		Translated: result.Translated,
		Notes:      result.Notes,
		Advisories: result.Advisories,
	})
}

// APITranslateMacro translates a macro file and returns the recording without
// loading it anywhere, which is what converting a library of them in one pass
// needs. It takes the file as an upload or as JSON, because a shell script
// and a program reach for different ones.
func (app *App) APITranslateMacro(c *gin.Context) {
	name, source, err := macroFileFromRequest(c)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	result := macroimport.Translate(source)
	// No session is in play, so the recording names no host. A caller that
	// replays it through a session gets that session's host; a caller writing
	// the file out fills the field in.
	workflow := macroWorkflow("", 0, name, result)
	if err := macroBlocksResolve(workflow); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, macroTranslationResponse{
		Name:       recordingNameForMacro(name),
		StepTotal:  len(workflow.Steps),
		Statements: result.Statements,
		Translated: result.Translated,
		Notes:      result.Notes,
		Advisories: result.Advisories,
		Workflow:   workflow,
	})
}

// macroFileFromRequest reads the macro from a multipart upload or from a JSON
// body carrying the same two things a file has: a name and its text.
func macroFileFromRequest(c *gin.Context) (string, string, error) {
	if file, err := c.FormFile("macro"); err == nil {
		reader, err := file.Open()
		if err != nil {
			return "", "", errors.New("the uploaded macro file could not be read")
		}
		defer reader.Close()
		if file.Size > maxMacroUploadBytes {
			return "", "", fmt.Errorf("macro file exceeds %d bytes", maxMacroUploadBytes)
		}
		payload, err := io.ReadAll(io.LimitReader(reader, maxMacroUploadBytes+1))
		if err != nil {
			return "", "", errors.New("the uploaded macro file could not be read")
		}
		if len(payload) > maxMacroUploadBytes {
			return "", "", fmt.Errorf("macro file exceeds %d bytes", maxMacroUploadBytes)
		}
		if len(payload) == 0 {
			return "", "", errors.New("the macro file is empty")
		}
		return file.Filename, string(payload), nil
	} else if !errors.Is(err, http.ErrMissingFile) && c.ContentType() == "multipart/form-data" {
		return "", "", errors.New("a macro file is required")
	}

	var body struct {
		Name    string `json:"name"`
		Content string `json:"content"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		return "", "", errors.New("send the macro as a file upload or as JSON with name and content")
	}
	if strings.TrimSpace(body.Content) == "" {
		return "", "", errors.New("the macro file is empty")
	}
	if len(body.Content) > maxMacroUploadBytes {
		return "", "", fmt.Errorf("macro file exceeds %d bytes", maxMacroUploadBytes)
	}
	return body.Name, body.Content, nil
}

// macroWorkflow wraps translated steps in the recording the rest of the app
// reads. The delays live on the steps that a wait in the macro asked for, so
// there is deliberately no every-step delay: inventing one would slow a flow
// the macro ran at full speed.
func macroWorkflow(host string, port int, sourceName string, result macroimport.Result) *WorkflowConfig {
	steps := result.Steps
	if steps == nil {
		steps = []session.WorkflowStep{}
	}
	return &WorkflowConfig{
		Host:        host,
		Port:        port,
		Name:        recordingNameForMacro(sourceName),
		Description: macroDescription(sourceName, result),
		Steps:       steps,
	}
}

// macroBlocksResolve checks that the branches and loops a translation
// produced close the way playback needs them to.
//
// It is the one thing about a translated recording nobody downstream can
// recover from: a step that did not translate is reported and a step that
// fails is logged, but a recording whose If has no EndIf is refused as a
// whole when it is played, and the operator is left with a file, a report
// that said the branch converted, and nothing that runs. The translator
// closes every block it opens; this is the assertion that says so at the
// point where the file would otherwise be handed over.
func macroBlocksResolve(workflow *WorkflowConfig) error {
	if workflow == nil || len(workflow.Steps) == 0 {
		return nil
	}
	if _, err := compileWorkflowSteps(workflow.Steps); err != nil {
		return fmt.Errorf("the translated recording's branches do not resolve, so it was not loaded: %v", err)
	}
	return nil
}

func macroDescription(sourceName string, result macroimport.Result) string {
	base := macroBaseName(sourceName)
	if base == "" {
		base = "a macro file"
	}
	if len(result.Notes) == 0 {
		return fmt.Sprintf("Imported from %s.", base)
	}
	lines := "lines"
	if len(result.Notes) == 1 {
		lines = "line"
	}
	// The count travels with the file. Somebody opening this recording in six
	// months should not have to guess whether it was a whole translation.
	return fmt.Sprintf("Imported from %s. %d %s of the macro did not translate and were left out.", base, len(result.Notes), lines)
}

// macroBaseName reduces whatever the client called the file to a bare name.
// A browser sends the name the operator picked, and a client sends whatever
// it likes; neither is allowed to introduce a path.
func macroBaseName(name string) string {
	name = strings.TrimSpace(name)
	if name == "" {
		return ""
	}
	name = strings.ReplaceAll(name, "\\", "/")
	name = path.Base(name)
	name = strings.TrimSpace(strings.Trim(name, "."))
	if name == "" || name == "/" {
		return ""
	}
	// Keep it to characters a filename and a piece of UI text can both carry.
	var b strings.Builder
	for _, r := range name {
		switch {
		case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z', r >= '0' && r <= '9':
			b.WriteRune(r)
		case r == '.', r == '-', r == '_', r == ' ':
			b.WriteRune(r)
		}
	}
	return strings.TrimSpace(b.String())
}

// recordingNameForMacro names the recording after the macro it came from, so
// a directory of converted files still says which is which.
func recordingNameForMacro(sourceName string) string {
	base := macroBaseName(sourceName)
	if base == "" {
		return "imported-macro.json"
	}
	if ext := filepath.Ext(base); ext != "" {
		base = strings.TrimSuffix(base, ext)
	}
	if base == "" {
		return "imported-macro.json"
	}
	return base + ".json"
}

// workflowTargetForSession answers with the host this session is pointed at,
// which is the one a translated recording will be replayed against.
func workflowTargetForSession(s *session.Session) (string, int) {
	var host string
	var port int
	withSessionLock(s, func() {
		host = s.TargetHost
		port = s.TargetPort
	})
	if port == 0 {
		port = 3270
	}
	return host, port
}
